src/QMCWaveFunctions/Fermion/DiracDeterminantBatched.cpp

```diff
@@ -197,6 +197,80 @@ typename DiracDeterminantBatched<DET_ENGINE>::Grad DiracDeterminantBatched<DET_E
   return g;
 }
 
+template<typename DET_ENGINE>
+void DiracDeterminantBatched<DET_ENGINE>::mw_evalGradWithSpin(
+    const RefVectorWithLeader<WaveFunctionComponent>& wfc_list,
+    const RefVectorWithLeader<ParticleSet>& p_list,
+    int iat,
+    std::vector<Grad>& grad_now,
+    std::vector<ComplexType>& spingrad_now) const
+{
+  assert(this == &wfc_list.getLeader());
+  auto& wfc_leader = wfc_list.getCastedLeader<DiracDeterminantBatched<DET_ENGINE>>();
+  ScopedTimer local_timer(RatioTimer);
+
+  const int nw = wfc_list.size();
+  RefVectorWithLeader<SPOSet> phi_list(*Phi);
+  std::vector<SPOSet::ValueVector> psi_values, lap_values, dspin_values;
+  std::vector<SPOSet::GradVector> grad_values;
+  RefVector<SPOSet::ValueVector> psi_v_list, lap_v_list, dspin_v_list;
+  RefVector<SPOSet::GradVector> grad_v_list;
+
+  phi_list.reserve(nw);
+  psi_values.reserve(nw);
+  grad_values.reserve(nw);
+  lap_values.reserve(nw);
+  dspin_values.reserve(nw);
+
+  psi_v_list.reserve(nw);
+  grad_v_list.reserve(nw);
+  lap_v_list.reserve(nw);
+  dspin_v_list.reserve(nw);
+
+  SPOSet::ValueVector val(wfc_leader.Phi->size());
+  SPOSet::GradVector grad(wfc_leader.Phi->size());
+  for (int iw = 0; iw < wfc_list.size(); iw++)
+  {
+    auto& det = wfc_list.getCastedElement<DiracDeterminantBatched<DET_ENGINE>>(iw);
+    phi_list.push_back(*det.Phi);
+
+    psi_values.emplace_back(val);
```

Comment:
So these are all getting constructed with the leader's copy of val? If that isn't incorrect i.e. these are only output variables why not just resize to nw.
Am I missing something?

Reply:
So here, val is just a temporary ValueVector. In order to have the right size of OrbitalSetSize (from the wfc_leader.Phi->size() call) to push into the std::vector<ValueVector> and then push back into the reference vector. 

All of this was done to just make it so that I can pass the RefVectors into the mw_evaluateVGLWithSpin call. I was having trouble getting things initialized to the right size using the constructors, so I turned to reserving the vector and RefVector sizes and then empalce_back. 

This method of getting the RefVectors of the right size was done in some unit tests, https://github.com/QMCPACK/qmcpack/blob/develop/src/QMCWaveFunctions/tests/test_einset.cpp#L520-L543

Reply:
Ok I see, it was far enough away that I missed it. I guess this could use a doc comment, in the spirit of self documentation consider:
``` c++
int num_orbitals = wfc_leader.Phi->size();
...
phi_list.emplace_back(num_orbitals);
```
Or If you wanted instead to move the construction outside of the for loop you could write the behavior of std::vector's constructors are pretty well known.
```
SPOSet::ValueVector sized_vector(wfc_leader.Phi->size());
std::vector<SPOSet::ValueVector> psi_values(wfc_list.size(),sized_vector);
```

Reply:
this has been fixed. this doesn't work with RefVector (which is std::vector<std::reference_wrapper<T>>, so I left those as they were

Reply:
we have a helper function `makeRefVector` in `type_traits/template_types.hpp`

Reply:
Oh ok great, I wasn't aware of that. Will fix

Reply:
done